File: Source/Utility/Common/Sequence.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { pColor } from "@rbxts/precomputed";

export function EvaluateNumberSequence(Sequence: NumberSequence, Time: number): number {
	Time = math.clamp(Time === 1 ? Time : Time % 1, 0, 1);

	if (Time === 0) {
		return Sequence.Keypoints[0].Value;
	} else if (Time === 1) {
		return Sequence.Keypoints[(Sequence.Keypoints.size() - 1)].Value;
	};

	for (const Index of $range(0, Sequence.Keypoints.size() - 1)) {
		const Current = Sequence.Keypoints[Index], Next = Sequence.Keypoints[(Index + 1)];

		if (Time >= Current.Time && Time < Next.Time) {
			const Alpha = (Time - Current.Time) / (Next.Time - Current.Time);

			return Current.Value + (Next.Value - Current.Value) * Alpha;
		};
	};

	return 0;
};

export function EvaluateColorSequence(Sequence: ColorSequence, Time: number): Color3 {
	Time = math.clamp(Time === 1 ? Time : Time % 1, 0, 1);

	if (Time === 0) {
		return Sequence.Keypoints[0].Value;

	} else if (Time === 1) {
		return Sequence.Keypoints[(Sequence.Keypoints.size() - 1)].Value;
	};

	for (const Index of $range(0, Sequence.Keypoints.size() - 1)) {
		const Current = Sequence.Keypoints[Index], Next = Sequence.Keypoints[(Index + 1)];

		if (Time >= Current.Time && Time < Next.Time) {
			const Alpha = (Time - Current.Time) / (Next.Time - Current.Time);

			return Current.Value.Lerp(Next.Value, Alpha);
		};
	};

	return pColor.White;
};

File: Output/Utility/Common/Sequence.lua
-- Compiled with roblox-ts v2.2.0
local TS = _G[script]
-- eslint-disable @typescript-eslint/no-explicit-any 
local pColor = TS.import(script, TS.getModule(script, "@rbxts", "precomputed").out).pColor
local function EvaluateNumberSequence(Sequence, Time)
	Time = math.clamp(if Time == 1 then Time else Time % 1, 0, 1)
	if Time == 0 then
		return Sequence.Keypoints[1].Value
	elseif Time == 1 then
		return Sequence.Keypoints[(#Sequence.Keypoints - 1) + 1].Value
	end
	for Index = 0, #Sequence.Keypoints - 1 do
		local Current = Sequence.Keypoints[Index + 1]
		local Next = Sequence.Keypoints[(Index + 1) + 1]
		if Time >= Current.Time and Time < Next.Time then
			local Alpha = (Time - Current.Time) / (Next.Time - Current.Time)
			return Current.Value + (Next.Value - Current.Value) * Alpha
		end
	end
	return 0
end
local function EvaluateColorSequence(Sequence, Time)
	Time = math.clamp(if Time == 1 then Time else Time % 1, 0, 1)
	if Time == 0 then
		return Sequence.Keypoints[1].Value
	elseif Time == 1 then
		return Sequence.Keypoints[(#Sequence.Keypoints - 1) + 1].Value
	end
	for Index = 0, #Sequence.Keypoints - 1 do
		local Current = Sequence.Keypoints[Index + 1]
		local Next = Sequence.Keypoints[(Index + 1) + 1]
		if Time >= Current.Time and Time < Next.Time then
			local Alpha = (Time - Current.Time) / (Next.Time - Current.Time)
			return Current.Value:Lerp(Next.Value, Alpha)
		end
	end
	return pColor.White
end
return {
	EvaluateNumberSequence = EvaluateNumberSequence,
	EvaluateColorSequence = EvaluateColorSequence,
}
